src/gtc/gtir_to_oir.py

```diff
@@ -13,11 +13,44 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 from dataclasses import dataclass, field
-from typing import Any, List
+from typing import Any, List, Set
 
 from eve import NodeTranslator
 from gtc import common, gtir, oir, utils
 from gtc.common import CartesianOffset, DataType, LogicalOperator, UnaryOperator
+from gtc.passes.oir_optimizations.utils import compute_fields_extents
+
+
+def validate_stencil_memory_accesses(node: oir.Stencil) -> oir.Stencil:
+    """Check that no memory races occur in GridTools backends.
+
+    Since this is required for GridTools backends, it's imposed on all backends
+    at the OIR level. This is similar to the check at the gtir level for read-with-offset
+    and writes, but more complete because it involves extent analysis, so it catches
+    indirect read-with-offset through temporaries.
+
+    """
+
+    def _writes(node: oir.Stencil) -> Set[str]:
+        result = set()
+        for left in node.iter_tree().if_isinstance(oir.AssignStmt).getattr("left"):
+            result |= left.iter_tree().if_isinstance(oir.FieldAccess).getattr("name").to_set()
+        return result
+
+    field_names = {decl.name for decl in node.params if isinstance(decl, oir.FieldDecl)}
```

Comment:
Would we not want to include temporaries here?

Reply:
Because temporaries are allowed to be written to and read with a horizontal offset in the same stencil, since they're block-local. This check is supposed to catch data arrays where there are race conditions. This only happens in input/output fields.